snuba/replacers/replacer_processor.py

```diff
@@ -0,0 +1,67 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Mapping, NamedTuple, Optional
+
+from snuba.datasets.schemas.tables import TableSchema, WritableTableSchema
+
+
+class ReplacementMessage(NamedTuple):
+    """
+    Represent a generic replacement message (version 2 in our protocol) that we
+    find on the replacement topic.
+    TODO: We should use codecs to encode/decode kafka replacements messages.
+    """
+
+    action_type: str  # This is a string to make this class agnostic to the dataset
+    data: Mapping[str, Any]
+
+
+@dataclass(frozen=True)
+class Replacement:
+    count_query_template: str
+    insert_query_template: str
+    query_args: Mapping[str, Any]
+    query_time_flags: Any
+
+
+class ReplacerProcessor(ABC):
+    """
+    Processes one message from the replacer topic into a data structure that contains
+    the query to apply the replacement.
+    Every dataset/storage that needs to implement replacements, needs to provide an
+    instance of this class that will be used by the ReplacementWorker.
+    """
+
+    def __init__(
+        self, write_schema: WritableTableSchema, read_schema: TableSchema
+    ) -> None:
+        self.__write_schema = write_schema
+        self.__read_schema = read_schema
+
+    @abstractmethod
+    def process_message(self, message: ReplacementMessage) -> Optional[Replacement]:
+        """
+        Processes one message from the topic.
+        """
+        raise NotImplementedError
+
+    def get_write_schema(self) -> WritableTableSchema:
+        return self.__write_schema
+
+    def get_read_schema(self) -> TableSchema:
+        return self.__write_schema
```

Comment:
Not self.__read_schema?

Reply:
Sorry I did not see this when I committed. You are right. This should be the read schema. It does not break anything right now because there is no replacement on any table where the schemas are not the same. Will fix